dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/BeanDeploymentArchiveVisitor.java

```diff
@@ -0,0 +1,125 @@
+/*******************************************************************************
+ * Copyright (c) 2024 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package com.ibm.ws.cdi.impl.weld;
+
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Stack;
+
+import com.ibm.websphere.ras.annotation.Trivial;
+import com.ibm.ws.cdi.CDIException;
+import com.ibm.ws.cdi.internal.interfaces.ArchiveType;
+import com.ibm.ws.cdi.internal.interfaces.WebSphereBeanDeploymentArchive;
+
+public class BeanDeploymentArchiveVisitor {
+
+    private final Stack<PeakableFilteredIterator> stack = new Stack<PeakableFilteredIterator>();
+
+    /**
+     * This method implements a depth first search over all BDAs (excluding Runtime Extensions) and their children
+     * to ensure children are scanned before parents.
+     *
+     * Since the dependency graph of BDAs is cyclic this algorithm can be nondeterministic in which order mutual
+     * dependencies are scanned.
+     *
+     * LOOP START
+     * Look at the current element from the iterator at the top of the stack, marking it visited:
+     * If it has valid (not visited or runtime extension) children place an iterator of its children on the top of its stack. GOTO LOOP START
```

Comment:
Why do runtime extensions need to be treated specially?

Reply:
I see one point where the old code was skipping runtime extensions, and I can see where you've added in "now scan the runtime extensions", but I can't see where runtime extensions got scanned in the old code.

Reply:
Looked at this again - I think the old logic was "don't scan children of runtime extensions", but I think the extension itself would still be scanned when it was encountered as a child of another archive?

In the new code, I think that runtime extensions won't be scanned until the end.